File: ShineMaps.Android/app/src/main/java/com/cokistudios/shinemaps/data/CSIDManager.kt
package com.cokistudios.shinemaps.data

import android.content.Context
import android.content.SharedPreferences
import android.util.Log
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONObject
import java.io.IOException
import java.util.UUID

data class CSIDUser(
    val id: String,
    val email: String,
    val name: String
) {
    val initial: String
        get() = if (name.isNotBlank()) name.take(1).uppercase() else email.take(1).uppercase()
}

class CSIDManager private constructor(context: Context) {

    private val prefs: SharedPreferences = context.getSharedPreferences("csid_auth_prefs", Context.MODE_PRIVATE)
    private val httpClient = OkHttpClient()

    private val baseUrl = "https://cmkumxprmmhuinxfppxl.supabase.co"
    private val anonKey = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6ImNta3VteHBybW1odWlueGZwcHhsIiwicm9sZSI6ImFub24iLCJpYXQiOjE3Nzc0OTkxNzEsImV4cCI6MjA5MzA3NTE3MX0.BNbSSxoObXMGpyin4-3udSM6ricoTO57Zaade5dTfxQ"

    var currentUser: CSIDUser? = null
        private set

    val isLoggedIn: Boolean
        get() = prefs.getString("csid_token", null) != null && currentUser != null

    val deviceHash: String
        get() {
            val existing = prefs.getString("csid_device_hash", null)
            if (existing != null) return existing
            val newHash = UUID.randomUUID().toString().replace("-", "")
            prefs.edit().putString("csid_device_hash", newHash).apply()
            return newHash
        }

    init {
        val savedId = prefs.getString("csid_user_id", null)
        val savedEmail = prefs.getString("csid_user_email", null)
        val savedName = prefs.getString("csid_user_name", null)
        if (!savedId.isNullOrBlank() && !savedEmail.isNullOrBlank()) {
            currentUser = CSIDUser(
                id = savedId,
                email = savedEmail,
                name = savedName ?: savedEmail.substringBefore("@")
            )
        }
    }

    suspend fun login(email: String, pass: String): Result<CSIDUser> = withContext(Dispatchers.IO) {
        try {
            val url = "$baseUrl/auth/v1/token?grant_type=password"
            val bodyJson = JSONObject().apply {
                put("email", email.trim())
                put("password", pass)
            }
            val request = Request.Builder()
                .url(url)
                .header("apikey", anonKey)
                .header("Content-Type", "application/json")
                .post(bodyJson.toString().toRequestBody("application/json".toMediaType()))
                .build()

            val response = httpClient.newCall(request).execute()
            val resStr = response.body?.string() ?: ""

            if (!response.isSuccessful) {
                val errorMsg = try {
                    val obj = JSONObject(resStr)
                    obj.optString("error_description", obj.optString("msg", "Error de autenticación"))
                } catch (e: Exception) {
                    "Error al conectar con CS ID"
                }
                return@withContext Result.failure(IOException(errorMsg))
            }

            val json = JSONObject(resStr)
            val accessToken = json.optString("access_token")
            val refreshToken = json.optString("refresh_token")
            val userObj = json.getJSONObject("user")
            val userId = userObj.optString("id")
            val userEmail = userObj.optString("email", email.trim())

            var userName = userEmail.substringBefore("@")
            val meta = userObj.optJSONObject("user_metadata")
            if (meta != null) {
                userName = meta.optString("full_name", meta.optString("name", userName))
            }

            val user = CSIDUser(userId, userEmail, userName)
            saveSession(accessToken, refreshToken, user)
            bindDeviceHash(userId, userEmail)

            Result.success(user)
        } catch (e: Exception) {
            Log.e("CSIDManager", "Login exception", e)
            Result.failure(e)
        }
    }

    suspend fun signUp(email: String, pass: String, name: String): Result<CSIDUser> = withContext(Dispatchers.IO) {
        try {
            val url = "$baseUrl/auth/v1/signup"
            val metaJson = JSONObject().apply {
                put("full_name", name.trim())
                put("name", name.trim())
                put("role", "user")
                put("company", "Coki Studios")
            }
            val bodyJson = JSONObject().apply {
                put("email", email.trim())
                put("password", pass)
                put("options", JSONObject().apply {
                    put("data", metaJson)
                    put("email_redirect_to", "https://cokistudios.github.io/coki-confirm.html")
                })
            }

            val request = Request.Builder()
                .url(url)
                .header("apikey", anonKey)
                .header("Content-Type", "application/json")
                .post(bodyJson.toString().toRequestBody("application/json".toMediaType()))
                .build()

            val response = httpClient.newCall(request).execute()
            val resStr = response.body?.string() ?: ""

            if (!response.isSuccessful) {
                val errorMsg = try {
                    val obj = JSONObject(resStr)
                    obj.optString("msg", obj.optString("message", "Error al crear cuenta CS ID"))
                } catch (e: Exception) {
                    "Error al crear cuenta CS ID"
                }
                return@withContext Result.failure(IOException(errorMsg))
            }

            // Auto-login after sign-up
            login(email, pass)
        } catch (e: Exception) {
            Log.e("CSIDManager", "SignUp exception", e)
            Result.failure(e)
        }
    }

    private fun saveSession(token: String, refresh: String, user: CSIDUser) {
        currentUser = user
        prefs.edit()
            .putString("csid_token", token)
            .putString("csid_refresh_token", refresh)
            .putString("csid_user_id", user.id)
            .putString("csid_user_email", user.email)
            .putString("csid_user_name", user.name)
            .apply()
    }

    fun logout() {
        currentUser = null
        prefs.edit()
            .remove("csid_token")
            .remove("csid_refresh_token")
            .remove("csid_user_id")
            .remove("csid_user_email")
            .remove("csid_user_name")
            .apply()
    }

    private fun bindDeviceHash(userId: String, email: String) {
        try {
            val url = "$baseUrl/rest/v1/user_device_hashes"
            val bodyJson = JSONObject().apply {
                put("device_hash", deviceHash)
                put("user_id", userId.lowercase())
                put("user_email", email)
            }
            val request = Request.Builder()
                .url(url)
                .header("apikey", anonKey)
                .header("Authorization", "Bearer $anonKey")
                .header("Content-Type", "application/json")
                .header("Prefer", "resolution=merge-duplicates")
                .post(bodyJson.toString().toRequestBody("application/json".toMediaType()))
                .build()
            httpClient.newCall(request).execute().close()
        } catch (e: Exception) {
            Log.w("CSIDManager", "bindDeviceHash failed: ${e.message}")
        }
    }

    companion object {
        @Volatile
        private var INSTANCE: CSIDManager? = null

        fun getInstance(context: Context): CSIDManager {
            return INSTANCE ?: synchronized(this) {
                INSTANCE ?: CSIDManager(context.applicationContext).also { INSTANCE = it }
            }
        }
    }
}

File: ShineMaps.iOS/CSIDManager.swift
import Foundation
internal import Combine

public class CSIDManager: ObservableObject {
    public static let shared = CSIDManager()

    private let baseUrl = "https://cmkumxprmmhuinxfppxl.supabase.co"
    private let anonKey = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6ImNta3VteHBybW1odWlueGZwcHhsIiwicm9sZSI6ImFub24iLCJpYXQiOjE3Nzc0OTkxNzEsImV4cCI6MjA5MzA3NTE3MX0.BNbSSxoObXMGpyin4-3udSM6ricoTO57Zaade5dTfxQ"

    @Published public var currentUser: CSIDUser? = nil
    @Published public var sessionToken: String? = nil

    public var isLoggedIn: Bool {
        currentUser != nil && sessionToken != nil
    }

    public var deviceHash: String {
        if let existing = UserDefaults.standard.string(forKey: "csid_device_hash") {
            return existing
        }
        let newHash = UUID().uuidString.replacingOccurrences(of: "-", with: "")
        UserDefaults.standard.set(newHash, forKey: "csid_device_hash")
        return newHash
    }

    private init() {
        self.sessionToken = UserDefaults.standard.string(forKey: "csid_session_token")
        if let savedId = UserDefaults.standard.string(forKey: "csid_user_id"),
           let savedEmail = UserDefaults.standard.string(forKey: "csid_user_email") {
            let savedName = UserDefaults.standard.string(forKey: "csid_user_name") ?? savedEmail.components(separatedBy: "@").first ?? "Usuario"
            self.currentUser = CSIDUser(id: savedId, email: savedEmail, name: savedName)
        }
    }

    // MARK: - Login
    public func login(email: String, pass: String) async throws -> CSIDUser {
        let url = URL(string: "\(baseUrl)/auth/v1/token?grant_type=password")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue(anonKey, forHTTPHeaderField: "apikey")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        let body: [String: Any] = [
            "email": email.trimmingCharacters(in: .whitespacesAndNewlines),
            "password": pass
        ]
        request.httpBody = try JSONSerialization.data(withJSONObject: body)

        let (data, response) = try await URLSession.shared.data(for: request)
        guard let httpResponse = response as? HTTPURLResponse else {
            throw URLError(.badServerResponse)
        }

        if !(200...299).contains(httpResponse.statusCode) {
            let errorMsg: String
            if let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] {
                errorMsg = (json["error_description"] as? String) ?? (json["msg"] as? String) ?? "Error al autenticar con CS ID"
            } else {
                errorMsg = "Error al autenticar (\(httpResponse.statusCode))"
            }
            throw NSError(domain: "CSIDManager", code: httpResponse.statusCode, userInfo: [NSLocalizedDescriptionKey: errorMsg])
        }

        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
              let token = json["access_token"] as? String,
              let userObj = json["user"] as? [String: Any],
              let userId = userObj["id"] as? String else {
            throw URLError(.cannotParseResponse)
        }

        let userEmail = (userObj["email"] as? String) ?? email
        var userName = userEmail.components(separatedBy: "@").first ?? "Usuario"

        if let metadata = userObj["user_metadata"] as? [String: Any] {
            if let fullName = metadata["full_name"] as? String, !fullName.isEmpty {
                userName = fullName
            } else if let name = metadata["name"] as? String, !name.isEmpty {
                userName = name
            }
        }

        let user = CSIDUser(id: userId, email: userEmail, name: userName)

        await MainActor.run {
            self.sessionToken = token
            self.currentUser = user
            UserDefaults.standard.set(token, forKey: "csid_session_token")
            UserDefaults.standard.set(userId, forKey: "csid_user_id")
            UserDefaults.standard.set(userEmail, forKey: "csid_user_email")
            UserDefaults.standard.set(userName, forKey: "csid_user_name")
        }

        Task.detached {
            await self.bindDeviceHash(userId: userId, email: userEmail)
        }

        return user
    }

    // MARK: - Sign Up
    public func signUp(email: String, pass: String, name: String) async throws -> CSIDUser {
        let url = URL(string: "\(baseUrl)/auth/v1/signup")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue(anonKey, forHTTPHeaderField: "apikey")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        let metadata: [String: Any] = [
            "full_name": name.trimmingCharacters(in: .whitespacesAndNewlines),
            "name": name.trimmingCharacters(in: .whitespacesAndNewlines),
            "role": "user",
            "company": "Coki Studios"
        ]

        let body: [String: Any] = [
            "email": email.trimmingCharacters(in: .whitespacesAndNewlines),
            "password": pass,
            "options": [
                "data": metadata,
                "email_redirect_to": "https://cokistudios.github.io/coki-confirm.html"
            ]
        ]
        request.httpBody = try JSONSerialization.data(withJSONObject: body)

        let (data, response) = try await URLSession.shared.data(for: request)
        guard let httpResponse = response as? HTTPURLResponse else {
            throw URLError(.badServerResponse)
        }

        if !(200...299).contains(httpResponse.statusCode) {
            let errorMsg: String
            if let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] {
                errorMsg = (json["msg"] as? String) ?? (json["message"] as? String) ?? "Error al registrar cuenta CS ID"
            } else {
                errorMsg = "Error al registrar (\(httpResponse.statusCode))"
            }
            throw NSError(domain: "CSIDManager", code: httpResponse.statusCode, userInfo: [NSLocalizedDescriptionKey: errorMsg])
        }

        // Auto login after successful signup
        return try await login(email: email, pass: pass)
    }

    // MARK: - Logout
    public func logout() {
        self.sessionToken = nil
        self.currentUser = nil
        UserDefaults.standard.removeObject(forKey: "csid_session_token")
        UserDefaults.standard.removeObject(forKey: "csid_user_id")
        UserDefaults.standard.removeObject(forKey: "csid_user_email")
        UserDefaults.standard.removeObject(forKey: "csid_user_name")
    }

    // MARK: - Device Hash Sync
    private func bindDeviceHash(userId: String, email: String) async {
        guard let url = URL(string: "\(baseUrl)/rest/v1/user_device_hashes") else { return }
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue(anonKey, forHTTPHeaderField: "apikey")
        request.setValue("Bearer \(anonKey)", forHTTPHeaderField: "Authorization")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("resolution=merge-duplicates", forHTTPHeaderField: "Prefer")

        let body: [String: Any] = [
            "device_hash": deviceHash,
            "user_id": userId.lowercased(),
            "user_email": email
        ]

        if let data = try? JSONSerialization.data(withJSONObject: body) {
            request.httpBody = data
            _ = try? await URLSession.shared.data(for: request)
        }
    }
}
